internal/datasource/config/config_writer.go

```diff
@@ -0,0 +1,195 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/service/config"
+
+	"github.com/google/uuid"
+	"github.com/nginx/agent/v3/api/grpc/instances"
+	"github.com/nginx/agent/v3/internal/client"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.7.0 -generate
+//counterfeiter:generate . ConfigWriterInterface
+type ConfigWriterInterface interface {
+	Write(filesUrl string, tenantID uuid.UUID) (err error)
+	Complete() (err error)
+}
+
+const (
+	cacheLocation = "/var/lib/nginx-agent/config/%v/cache.json"
+)
+
+type (
+	Client struct {
+		Timeout time.Duration
+	}
+
+	ConfigWriterParameters struct {
+		configClient client.HttpConfigClientInterface
+		Client       Client
+		cachePath    string
+	}
+
+	ConfigWriter struct {
+		configClient       client.HttpConfigClientInterface
+		previouseFileCache FileCache
+		currentFileCache   FileCache
+		cachePath          string
+		dataplaneConfig    config.DataplaneConfig
+	}
+
+	// map of files with filepath as key
+	FileCache = map[string]*instances.File
+)
+
+func NewConfigWriter(configWriterParameters *ConfigWriterParameters, instanceId string) *ConfigWriter {
+	if configWriterParameters.cachePath == "" {
```

Comment:
I think it might be a good idea to add a `nil` check for `configWriterParameters`, as otherwise this is can be a `nil` pointer dereference.

Reply:
should we default the values in the configWriterParameters so it is never nil and the you can remove the cachePath default value from here? Whichever one is better (this or whatever I suggested)